### DISCUSSION TITLE
Can I improve me meross connection?

### DISCUSSION THREAD

**Discussion by userMak** [krahabb/meross_lan#192] [General]
Hi
It's over 6-8 months that I have connected all my meross light switches though mqtt and in general I am happy with the integration.
However there are times that 1-2 and a few times all my devices are disconnected randomly with no reason. I have to connect them with npx meross setup.

right now in the configuration of each device I have the following
Is there something I can change in this configuration to make the connection more stable? 

![image](https://user-images.githubusercontent.com/55662933/169574692-1d0570a6-4669-4449-9759-0d3777e598c0.png)

**Comment by krahabb** [comment]
@userMak ,
Do you really need to go over npx setup again? That means the devices loose their internal configuration which is very unlikely (unless their flash-rom is faulted)
Does the mosquitto log say anything? 

There could be another reason:
if your devices are the wall mount switches (like mss550 or so) they could be reset by inadvertently press the button for more than 5 seconds. In general I guess many devices could be reset this way or by multiple back-to-back toggles at the switch.
The light bulbs for instance can be reset by switching 3 times in a row (or something like that - I don't remember very well)...
You could try investigate if this is the case with your devices

**Reply by userMak** [reply]
> 

Yes, I have to use npx setup. I haven't looked in the logs, but the last time I was in a quiet room and I heard "out of the blue" my wall switch clicking and after 2-3 seconds started to blink red. There was no physical contact or anything. I only have 7 wall switches like mss550), and as I said once every a month or 2 months one of them disconnects. There are 2 times in a year that I found them all blinking red. I will try to check the logs next time and report it here. If I have understand correct there is nothing I could change in the above setup screen it could help keep their connection.

**Comment by userMak** [comment]
@krahabb 
Hi
I was "lucky" and just now I heard one of the switches to click and disconnect for a while. this time it was reconnected after a few minutes
I saw it blinking red but it did reconnected.
the logs are the following. is this normal? 



> 1653252080: Saving in-memory database to /data/mosquitto.db.
1653253881: Saving in-memory database to /data/mosquitto.db.
1653255682: Saving in-memory database to /data/mosquitto.db.
1653257483: Saving in-memory database to /data/mosquitto.db.
1653257908: Client fmware:1812013756539529087334298f18976e_GgtDnkBTP9N8kB0N has exceeded timeout, disconnecting.
1653257923: New connection from 192.168.1.50 on port 8883.
[01:18:44] INFO: [INFO] found 34:29:8f:18:97:6e on Home Assistant
1653257924: New client connected from 192.168.1.50 as fmware:1812013756539529087334298f18976e_GgtDnkBTP9N8kB0N (p1, c1, k30, u'34:29:8f:18:97:6e').
1653257937: New connection from 192.168.1.50 on port 8883.
1653257937: Client fmware:1812013756539529087334298f18976e_GgtDnkBTP9N8kB0N already connected, closing old connection.
1653257937: New client connected from 192.168.1.50 as fmware:1812013756539529087334298f18976e_GgtDnkBTP9N8kB0N (p1, c1, k30, u'34:29:8f:18:97:6e').

**Reply by krahabb** [reply]
According to mosquitto documentation that is because of the 'keep alive timeout'.
Normally the device should send a sort of MQTT PING to the broker in order to keep the connection alive, if the broker doesn't receive any in the configured time period it will disconnect the client (like the log says).
Now, the issue shouldnt happen since the timeout is generally managed by the device itself.
the device tells the broker which timeout to use and then it should respect this. For example, the device can tell the broker: I'm using 60 seconds timeout and so the broker will just wait for this sort of PING coming from the device. Mosquitto usually waits at least 1.5 X timeout (so it waits 90 seconds in the example) before disconnecting the device should it not send the PING.
At least this is how I interpret the documentation I've found around.

Now, I don't know for sure which timeout the device should use, it's not a standard so the meross implementation could have chosen any but it normally works: to be honest I use my devices not so often so I can't really tell if they disconnect or not but when I use them they're always online.
The issue could lie in the device firmware or in the mosquitto implementation. You could try get deeper in the mosquitto documentation to see if there's any configuration parameter which could help or worsen the situation. Also, to see if the devices send those PINGs I guess you could raise the mosquitto log level in order to log these messages and see if they get sent by the device and the time period between them.

As a very patchy fix for this issue (if not any other) you could setup a periodic automation in HA to schedule every so often (1 minute - 5 minutes or so, depending if you can get which timeout the device should have). In this automation you could just send a switch.turn_on (or turn_off) to all of the DND light entities of the meross and see if the disconnections disappear.

Also, I've found this by just googling around: https://github.com/eclipse/mosquitto/issues/2117#issuecomment-969360753 so you maybe can explicitly configure mosquitto to allow for no keep_alive checks and to not disconnect such clients

**Reply by userMak** [reply]
@krahabb 

do you think that this will work?
this is my config.yaml and I added the max_keepalive: 0 line at the end

> logins: []
anonymous: true
customize:
  active: true
  folder: mosquitto
certfile: server.crt
keyfile: server.key
require_certificate: false
cafile: ca.crt
max_keepalive: 0

**Reply by krahabb** [reply]
not sure though but the idea is to add the maxkeepalive set to 0 according to the thread I've found.
Only time will tell (unless mosquitto raises new bugs!)

**Reply by userMak** [reply]
I added yesterday evening the max_keepalive: 0 line, but I don't think it is working. (unfortunately) 

> [09:03:41] INFO: [INFO] found 34:29:8f:18:61:d1 on Home Assistant
1653372221: New client connected from 192.168.1.48 as fmware:1811269316494129086734298f1861d1_wqcAdzyglw9s4S7p (p1, c1, k30, u'34:29:8f:18:61:d1').
1653372543: Client fmware:1812011610829529087334298f1896a5_DVyxLOd21ZK7pEYo has exceeded timeout, disconnecting.
1653372552: New connection from 192.168.1.49 on port 8883.
[09:09:13] INFO: [INFO] found 34:29:8f:18:96:a5 on Home Assistant
1653372553: New client connected from 192.168.1.49 as fmware:1812011610829529087334298f1896a5_DVyxLOd21ZK7pEYo (p1, c1, k30, u'34:29:8f:18:96:a5').
1653372843: Client fmware:1812016823740829087334298f18972e_KadgedVyzAsOUSpF has exceeded timeout, disconnecting.
1653372852: New connection from 192.168.1.51 on port 8883.
[09:14:12] INFO: [INFO] found 34:29:8f:18:97:2e on Home Assistant
1653372852: New client connected from 192.168.1.51 as fmware:1812016823740829087334298f18972e_KadgedVyzAsOUSpF (p1, c1, k30, u'34:29:8f:18:97:2e').

**Reply by krahabb** [reply]
I can't really tell...from what I've seen on the mosquitto threads this is an hell of an issue since it is not clear how this configuration should apply to the different MQTT protocol versions. Also, it is not clear how mosquitto effectively manages this.
So we could be in a situation where, depending on which version of mosquitto you are using, this is not applied to v3 clients (I guess meross devices are not MQTTv5) or it is applied and the behaviour is unclear.
As a different approach you could try setting a very high value (maybe 65535 is the maximum but I suggest you to check the mosquitto man).

If not any, I'd suggest you again to setup a scheduled automation which periodically sends an mqtt message to the devices (you could use the meross_lan provided service and issue a command 'GET' for namespace 'Appliance.System.All'

**Reply by userMak** [reply]
> f not any, I'd suggest you again to setup a scheduled automation which periodically sends an mqtt message to the devices (you could use the meross_lan provided service and issue a command 'GET' for namespace 'Appliance.System.All'

I am awful with mqtt in general, and regarding the automations. If you got time can you please give me an example how I could automate it? I don't understand how to code the above.

**Comment by krahabb** [comment]
Yep,
You have to setup an automation which runs on a scheduled interval so you have to use the 'time_pattern' trigger type. Then put no conditions at all and just add a sequence of actions for each device: each action should just send a valid mqtt message to a single device by using the 'meross_lan.request' service call.
This is a snippet of an automation pinging 2 devices. Just put in your devices UUIDs (you can grab the UUID from the configuration panel of each meross_lan integration/device)
```yaml
alias: PingMyMeross
description: ''
trigger:
  - platform: time_pattern
    minutes: /5
condition: []
action:
  - service: meross_lan.request
    data:
      namespace: Appliance.System.All
      device_id: 2034188094237526280848e1e93aff26
      method: GET
  - service: meross_lan.request
    data:
      namespace: Appliance.System.All
      device_id: 2034188094237526280848e1e93aff27
      method: GET
mode: single
```
This automation is setup to run every 5 minutes and query 2 devices. I suggest you to start with a relatively high time pattern and eventually reduce it if it happens your devices still get timeout disconnected.
If you programmed your devices with a non empty key, then you'll have to also add it to each service call (this is not strictly needed though since a wrong key would anyway ping the device, it will just reply with an error but that's not an issue):
```yaml
  - service: meross_lan.request
    data:
      namespace: Appliance.System.All
      device_id: 2034188094237526280848e1e93aff27
      method: GET
      key: supersecretkey
```

**Reply by userMak** [reply]
Thanks! I will try it and let you know.

**Comment by userMak** [comment]
Just for your information. I added the above automation (every 30 minutes) however the result is the same. They haven't completely disconnected (yet) but the pattern is the same

> 1654028676: New connection from 192.168.1.51 on port 8883.
[23:24:36] INFO: [INFO] found 34:29:8f:18:97:2e on Home Assistant
1654028676: New client connected from 192.168.1.51 as fmware:1812016823740829087334298f18972e_KadgedVyzAsOUSpF (p1, c1, k30, u'34:29:8f:18:97:2e').
1654029545: Saving in-memory database to /data/mosquitto.db.
1654031346: Saving in-memory database to /data/mosquitto.db.
1654033147: Saving in-memory database to /data/mosquitto.db.
1654034948: Saving in-memory database to /data/mosquitto.db.
1654036749: Saving in-memory database to /data/mosquitto.db.
1654038550: Saving in-memory database to /data/mosquitto.db.
1654040351: Saving in-memory database to /data/mosquitto.db.
1654042152: Saving in-memory database to /data/mosquitto.db.
1654043953: Saving in-memory database to /data/mosquitto.db.
1654045754: Saving in-memory database to /data/mosquitto.db.
1654047555: Saving in-memory database to /data/mosquitto.db.
1654049356: Saving in-memory database to /data/mosquitto.db.
1654051157: Saving in-memory database to /data/mosquitto.db.
1654052958: Saving in-memory database to /data/mosquitto.db.
1654054759: Saving in-memory database to /data/mosquitto.db.
1654056560: Saving in-memory database to /data/mosquitto.db.
1654058361: Saving in-memory database to /data/mosquitto.db.
1654060162: Saving in-memory database to /data/mosquitto.db.
1654061729: Client fmware:1811264836268229086734298f1862d0_0LKxARsufXxIZ9qn has exceeded timeout, disconnecting.
1654061736: New connection from 192.168.1.52 on port 8883.
[08:35:37] INFO: [INFO] found 34:29:8f:18:62:d0 on Home Assistant
1654061737: New client connected from 192.168.1.52 as fmware:1811264836268229086734298f1862d0_0LKxARsufXxIZ9qn (p1, c1, k30, u'34:29:8f:18:62:d0').

**Reply by krahabb** [reply]
I see,
I expect the 30 minutes period does not solve the issue since meross_lan itself has an 'heartbeat' which automatically sends an MQTT query to any device after 5 minutes of 'inactivity' so, this means that a device MQTT 'channel' doesn't idle for more than 5 minutes already.
If mosquitto is disconnecting for the idle timeout that would likely be under 5 minutes so you really need to run your automation more often than that.

**Reply by userMak** [reply]
Thanks, to be honest I am not planning to even try to run this automation for less than 5 minutes. I am (was) looking/hoping  for something more 'elegant', but in every case thanks for your time & help.
